contracts-periphery/src/UmbraBatchSend.sol

```diff
@@ -24,54 +26,83 @@ contract UmbraBatchSend {
     bytes32 ciphertext;
   }
 
+  mapping(address => uint256) internal totalTransferAmountPerToken;
+  struct TransferSummary {
+    uint256 amount;
+    address tokenAddr;
+  }
+
   error ValueMismatch();
+  error TransferAmountMismatch();
   event BatchSendExecuted(address indexed sender);
 
   constructor(IUmbra _umbra) {
     umbra = _umbra;
   }
 
-  function batchSendEth(uint256 _tollCommitment, SendEth[] calldata _params) external payable {
-    if(msg.value != _sumValueFromEthSends(_tollCommitment, _params)) revert ValueMismatch();
+  function batchSendEth(uint256 _tollCommitment, SendEth[] calldata _params) public payable {
+    _checkValueFromSends(msg.value, _tollCommitment, _params, new SendToken[](0));
     _batchSendEth(_tollCommitment, _params);
     emit BatchSendExecuted(msg.sender);
   }
 
-  function batchSendTokens(uint256 _tollCommitment, SendToken[] calldata _params) external payable {
-    if(msg.value != _tollCommitment * _params.length) revert ValueMismatch();
-    _batchSendTokens(_tollCommitment, _params);
+  function batchSendTokens(
+    uint256 _tollCommitment,
+    SendToken[] calldata _params,
+    TransferSummary[] calldata _transferSummary
+  ) public payable {
+    _checkValueFromSends(msg.value, _tollCommitment, new SendEth[](0), _params);
+
+    for (uint256 i = 0; i < _params.length; i = _uncheckedIncrement(i)) {
+      // Used later to validate total amounts are correctly provided
+      totalTransferAmountPerToken[_params[i].tokenAddr] += _params[i].amount;
+    }
+
+    uint256 _summaryLength = _transferSummary.length;
+    for (uint256 i = 0; i < _summaryLength; i = _uncheckedIncrement(i)) {
+      IERC20 _token = IERC20(address(_transferSummary[i].tokenAddr));
+
+      if (totalTransferAmountPerToken[_transferSummary[i].tokenAddr] != _transferSummary[i].amount) {
+        revert TransferAmountMismatch();
+      }
+
+      IERC20(_token).safeTransferFrom(msg.sender, address(this), _transferSummary[i].amount);
+    }
+    _batchSendTokens(_tollCommitment, _params, _transferSummary);
+
     emit BatchSendExecuted(msg.sender);
   }
 
   function batchSend(
     uint256 _tollCommitment,
     SendEth[] calldata _ethParams,
-    SendToken[] calldata _tokenParams
+    SendToken[] calldata _tokenParams,
+    TransferSummary[] calldata _transferSummary
   ) external payable {
-    uint256 _totalAmount = _sumValueFromEthSends(_tollCommitment, _ethParams) + (_tollCommitment * _tokenParams.length);
-    if(msg.value != _totalAmount) revert ValueMismatch();
-
-    _batchSendEth(_tollCommitment, _ethParams);
-    _batchSendTokens(_tollCommitment, _tokenParams);
+    batchSendEth(_tollCommitment, _ethParams);
+    batchSendTokens(_tollCommitment, _tokenParams, _transferSummary);
```

Comment:
This duplicates the `_checkValueFromSends` validation right? I think we can:
- Make both `function batchSend*` methods external instead of public
- Move the event emissions into the internal `function _batchSend*` methods
- In `batchSendTokens` method, it should mirror the `batchSendEth` method in that it only (1) checks the values, and (2) calls the internal version. So the for loops in `batchSendTokens` can be internal to `_batchSendTokens` instead
- All `_checkValueFromSends` calls are in the external methods, and the internal ones handle everything else

I think that should work?

Reply:
It does. Implemented here: https://github.com/ScopeLift/umbra-protocol/pull/351/commits/fec577097bc55395bf33e0a7079845e3323a95df